### PR TITLE
Server basically complete

### PR DESCRIPTION
**Pull request by networking-sorceress**
The only thing left to do now is to implement the encryption functions, which will just be system function calls.

The client is going to be interesting. Backend-wise, it will be fairly simple. The front-end is the challenging part though. Haven't looked for python3 ncurses libs yet, but I bet they're just as cancer as the C ones.

### CODE REVIEW

File: src/server.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        self.clientLock = threading.RLock();
+
+        # Initialize global message queue
+        self.globalMessageQueue = MessageQueue();
+
+    def bind(self, port):
+        # Binds server socket to all interfaces on given port. Expects UInt16.
+
+        # Bind server to all interfaces
+        self.serverSocket.bind(("0.0.0.0",port));
```
Comment (Aqcurate) [inline src/server.py:45]: Manage the socket.error exception?

File: src/server.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                handler = Client(client, uname, pubkey, self);
+
+                # Add new handler to client list
+                with self.clientLock:
+                    self.clients.append(handler);
+
+                # Run handler thread
+                handler.start();
+                    
+
```
Comment (Aqcurate) [inline src/server.py:116]: Add a way to close the server connection? def exit(self):

File: main.py [hunk 3]
```diff
@@ -17,17 +17,16 @@
 # End module imports
 
 # Begin user imports
-#import src.RequestHandler
+from src.server import ChatdServer as Server
 # End user imports
 
 # main
```
Comment (Aqcurate) [inline main.py:18]: Wrap everything in if \_\_name__ = '\_\_main__' to make the code only run when it is executed as a program

File: src/server.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+# Begin class
+class ChatdServer(object):
+    # ChatdServer -- Manage chat threads and socket objects to facilitate
+    # communication between all clients
+
+    def __init__(self):
+        # Takes no arguments, returns new ChatdServer object
+
+        # Initialize server socket
+        self.serverSocket = socket.socket();
```
Comment (EscapingGithub) [inline src/server.py]: Pick either snake case or camel case

File: src/server.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+#
+# server.py
+# Created by Stephen Brimhall on 1/23/17
+# Copyright (c) SST CTF and Stephen Brimhall. All Rights Reserved
+#
+# This file implements the `server` class for the chatd instant messaging system
+
+# Begin system imports
+import sys;
```
Comment (Aqcurate) [inline src/server.py]: Is the system import ever used?

File: src/connection.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        threading.Thread.__init__(self);
+
+    def run(self):
+        # Implement listener
+
+        jsonDecoder = json.JSONDecoder();
+
+        while True:
+            # Block until data received
+            data = self.client.recv(4096);
```
Comment (Aqcurate) [inline src/connection.py:66]: Do you plan on dealing with data >4096 bytes?

File: src/queue.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# Implements a simple threadsafe queue object.
+#
+
+# Begin system imports
+from threading import RLock as Lock;
+# End system imports
+
+# Queue class
+
+class Queue(object):
```
Comment (Aqcurate) [inline src/queue.py:17]: I'm assuming the Queue system is unfinished?

**Review by Aqcurate — DISMISSED** [review]
REDACTED

**Review by Aqcurate — COMMENTED** [review]
Alright, here are my pedant complaints. Feel free to ignore them all. Everything works as is.

Make sure you remove \_\_pycache__.

For future reference, variables should use snake_case, following similar method naming conventions. Semicolons should not be used. (It would be a pain to change that now though, so leave it as is.)

I'm going to assume the Queue system and encryption system is unimplemented, so I only gave it a cursory glance.

I'm not too too familiar with threading and sockets, so I'll leave it to you to handle exceptions.

I didn't get a chance to run the code, but it looks fine. Just look over my comments, and I'll merge if there are no further changes.

File: src/server.py (hunk 1 above)
Comment (networking-sorceress) [reply src/server.py:45]: Actually, for this one, I want it to crash if it fails. That way the person who runs it knows there's already something running on the port.

File: src/connection.py (hunk 6 above)
Comment (networking-sorceress) [reply src/connection.py:66]: Yes. In the client. By saying you can't send more than 2,000 characters, leaving the other 2k for metadata.

File: src/queue.py (hunk 7 above)
Comment (networking-sorceress) [reply src/queue.py:17]: Pretty much. I am just using it as a way to organize threadsafe lists for now, I may implement some safer way of managing it later but for now good practice will do.

File: src/server.py (hunk 5 above)
Comment (networking-sorceress) [reply src/server.py]: No, I was importing it at first because I thought I might. I'm gonna go through and remove all the unused imports.

**Comment by networking-sorceress** [comment]
Changes are made.

File: main.py [hunk 8]
```diff
@@ -10,24 +10,17 @@
...
-
-# Bind socket to localhost port 
-sock.bind(("localhost",20039));
-
-newsock,_ = sock.accept();
+# Initialize the server
+serv = Server();
 
+# Bind the server
+serv.bind(10001);
```
Comment (otakaran) [inline main.py:23]: Has the port been cleared? What are you going to do about cross-compatability, or are you just hoping for the best? Is there a way to automatically search and open a port on a server? Having conflicts here would not be good...

File: src/connection.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import src.crypt;
+from src.sender import Sender;
+# End local imports
+
+# Static vars
+QUIT_MSGS = [
+    ":q",
+    ":d",
+    ":quit",
+    ":disconnect"
```
Comment (otakaran) [inline src/connection.py:28]: Add 'exit'
_Assuming there would be no conflict_

File: src/connection.py (hunk 6 above)
Comment (otakaran) [reply src/connection.py:66]: This seems like an unnecessary limitation... Could there be a convenient config option that admins can change. Also, what kind of message will the user receive if data > 2,000 chars?

File: src/server.py (hunk 1 above)
Comment (otakaran) [reply src/server.py:45]: How would the user know the crash was caused by a port config problem instead os something else. At least give the user a print out error!

File: src/connection.py (hunk 6 above)
Comment (networking-sorceress) [reply src/connection.py:66]: This is outside of the scope of this project. This is strictly the server side. Those messages will be handled on the client side.

**Review by Aqcurate — APPROVED** [review]
Alright, everything looks good!

**Pull request merged**